pytests/helpers/timeline.md

```diff
@@ -0,0 +1,313 @@
+```py
+something = timeline.mark('something')
+something >> Event('stopped', ANY)
+```
+a timeline like this:
+```py
+('Event', 'stopped', {'reason': 'breakpoint', 'threadId': 1})
+('Mark', 'something')
+```
+will not realize `something`, but this one will:
+```py
+('Event', 'stopped', {'reason': 'breakpoint', 'threadId': 1})
+('Mark', 'something')
+('Event', 'stopped', {'reason': 'breakpoint', 'threadId': 2})
+```
+
+Conversely, `(X >> O)` is an expectation that is realized at the first occurrence at  which `X` is realized, and which **precedes** `O` (and, again, it cannot be `O` itself!). So:
+```py
+something = timeline.mark('something')
+Event('stopped', ANY) >> something
+```
+will be realized at `something`, but only if the timeline already had an event "stopped" at that moment - since timelines only grow into the future, it is impossible for an event necessary to realize this expectation to appear after `mark()` was invoked.
+
+In practice, this is most often used with requests and responses; for example, to describe an event that should occur between a specific request and its response:
+```py
+initialize = debug_session.send_request('initialize', {'adapterID': 'test'})
+initialize_response = initialize_request.wait_for_response()
+assert (
+    initialize
+    >>
+    Event('initialized', {})
+    >>
+    initialize_response
+) in debug_session.timeline
+```
+Another useful pattern is `>>` combined with `wait_until`:
+```
+initialize = debug_session.send_request('initialize', {'adapterID': 'test'})
+initialized = debug_session.wait_until(Event('initialized'))
+assert (
+    initialize
+    >>
+    Event('output', ANY.dict_with({'category': 'telemetry'}))
+    >>
+    initialized
+) in debug_session.timeline
+```
+
+### Conjuction (`&`)
+
+When two expectations are conjuncted: `(A & B)` - the resulting expectation is realized at the occurrence at which `A` and `B` are both realized, regardless of their relative order. Thus:
+```py
+Event('stopped', ANY) & Event('continued', ANY)
+```
+this expectation will be realized in timeline:
+```py
+('Event', 'continued', {'threadId': 1})
+('Event', 'thread', {'reason': 'exited', 'threadId': 1})
+('Event', 'stopped', {'reason': 'breakpoint', 'threadId': 2})
+```
+but also in a differently ordered timeline:
+```py
+('Event', 'stopped', {'reason': 'breakpoint', 'threadId': 2})
+('Event', 'continued', {'threadId': 1})
+('Event', 'thread', {'reason': 'exited', 'threadId': 1})
+```
+This is most commonly used as seen above, with related events where their relative ordering is unspecified (usually also framed by `>>` to narrow it down to a specific request/response pair). It can also be used to concurrently send multiple requests, and wait until they all got their responses:
+```py
+pause1 = debug_session.send_request('pause', {'threadId': '1'})
+pause2 = debug_session.send_request('pause', {'threadId': '2'})
+debug_session.wait_until(Request(pause1) & Request(pause2))
```

Comment:
>Request(pause1) & Request(pause2) [](start = 25, length = 33)

Should this be just pause1 & pause2

Reply:
It should actually be `Response(pause1) & Response(pause2)`.